### PR TITLE
[GSOC 2025] 3D FEM

### PR DESCRIPTION
**Pull request by Rishab87**
# Description

This PR will add a support for 3D FEM, 3D meshing along with a mesh generator with support for prismatic and cylindrical geometry

Fixes # (issue)

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #)

# Important checks:

Please confirm the following before marking the PR as ready for review:
- [x] No style issues: `nox -s pre-commit`
- [x] All tests pass: `nox -s tests`
- [x] The documentation builds: `nox -s doctests`
- [x] Code is commented for hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -42,6 +42,7 @@ dependencies = [
     "posthog",
     "pyyaml",
     "platformdirs",
+    "pygmsh>=7.1.11"
```
Comment (kratman) [inline pyproject.toml:45]: Since this is GPLv3, is there a different package that can be used instead?

Comment (Rishab87) [reply pyproject.toml:45]: How about PyMesh, can we use that? It has MIT license I think.

**Comment by ig-raiderler** [comment]
The documentation in this repository is excellent! Very clear and comprehensive.

File: pyproject.toml (hunk 1 above)
Comment (rtimms) [reply pyproject.toml]: looks like pymesh hasn't been updated in a long time. we already use scikit-fem for 2D finite elements and meshing -- do they support 3d?

**Review by rtimms — COMMENTED** [review]
Looking good so far. Next steps are:

1. Add an example of solving the 3D heat eqn in a jelly and prismatic domain, with different boundary conditions and a prescribed heat souce
2. Add tests for the 3d finite volume methods (make sure things are discredited as expected, check simple problems are solved correctly). To help with this you can use the "method of manufactured solutions".

Once the mesh and method are tested for simple general equations we can start on the battery-specific parts of the project in a separate PR

File: pyproject.toml (hunk 1 above)
Comment (Rishab87) [reply pyproject.toml]: yeah they do, migrating to sickit-fem then

**Comment by aabills** [comment]
Hey Rishab, yeah Ghost nodes are a real pain. Let me wrap mine up this week and I'll get back to you ASAP

**Comment by Rishab87** [comment]
thanks @aabills, in meantime I'll try to fix these errors myself.

**Comment by Rishab87** [comment]
I've removed spiral added heat diffusion examples everything seems to be working now, would love to get some feedback on this

File: docs/source/api/meshes/index.rst [hunk 2]
```diff
@@ -7,3 +7,4 @@ Meshes
   zero_dimensional_submeshes
   one_dimensional_submeshes
   two_dimensional_submeshes
+  scikit_fem_submeshes_3d
```
Comment (rtimms) [inline docs/source/api/meshes/index.rst]: minor, but can we call this "three_dimensional_submeshes" for consistency with the other names

File: examples/scripts/3d_equation_cylindric.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    SpatialVariable,
+)
+
+R = 0.5
+H = 1.0
+alpha = 0.1
+
+model = BaseModel()
+
+x = SpatialVariable("x", ["current collector"], coord_sys="cartesian", direction="x")
```
Comment (rtimms) [inline examples/scripts/3d_equation_cylindric.py]: Maybe I'm misunderstanding, but I'm confused as to why we aren't using cylindrical variables here? I was expecting to do something like solve the heat equation in cylindrical coordinates in 3D.

File: examples/scripts/3d_equation_cylindric.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    }
+}
+
+submesh_types = {
+    "current collector": pybamm.ScikitFemGenerator3D(
+        geom_type="cylinder", h=0.1, radius=R, height=H
+    )
+}
+
+var_pts = {x: None, y: None, z: None}
```
Comment (rtimms) [inline examples/scripts/3d_equation_cylindric.py]: why are the nodes `None`?

File: src/pybamm/expression_tree/unary_operators.py [hunk 5]
```diff
@@ -905,8 +908,17 @@ def __init__(self, child, region="entire"):
             name += "negative tab"
         elif region == "positive tab":
             name += "positive tab"
+        elif region == "top":
```
Comment (rtimms) [inline src/pybamm/expression_tree/unary_operators.py]: we will need a better way to generalise this to other domains. what do top/bottom/left/right mean in 3D?

File: src/pybamm/meshes/scikit_fem_submeshes_3d.py [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+    ----------
+    geom_type : str
+        Type of geometry to generate ('box', 'cylinder', 'spiral')
+    **gen_params : dict
+        Geometry-specific parameters
+    """
+
+    def __init__(self, geom_type, **gen_params):
+        super().__init__(ScikitFemSubMesh3D)
+        self.geom_type = geom_type
```
Comment (rtimms) [inline src/pybamm/meshes/scikit_fem_submeshes_3d.py:32]: If this can only be a certain type, can we add some checks here? In the first instance, I was expecting to be able to mesh a cylinder (idealised jelly roll) and a rectangular prism (idealised pouch or prismatic cell).

File: src/pybamm/meshes/scikit_fem_submeshes_3d.py [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+            x_key = next(k for k in lims if k.name == "x")
+            y_key = next(k for k in lims if k.name == "y")
+            z_key = next(k for k in lims if k.name == "z")
+            x_lim = tuple(lims[x_key].values())
+            y_lim = tuple(lims[y_key].values())
+            z_lim = tuple(lims[z_key].values())
+            mesh = self._make_box_mesh(x_lim, y_lim, z_lim, h)
+
+        elif self.geom_type == "cylinder":
+            radius = self.gen_params.get("radius", 0.4)
```
Comment (rtimms) [inline src/pybamm/meshes/scikit_fem_submeshes_3d.py]: I don't think these should have defaults. It might lead to incorrect geometries being created silently by accident.

File: tests/shared.py [hunk 8]
```diff
@@ -116,6 +116,152 @@ def get_mesh_for_testing(
...
+    generator_params = {"h": 0.2}
+    generator_params.update(geom_params)
+
+    generator = pybamm.ScikitFemGenerator3D(geom_type, **generator_params)
+
+    submesh_types = {
+        "current collector": generator,
+    }
+
+    var_pts = {x: xpts, y: ypts, z: zpts}
```
Comment (rtimms) [inline tests/shared.py]: can we specify the points or do they change based on how the mesh is generated?

File: tests/unit/test_meshes/test_scikit_fem_submesh_3d.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+        assert submesh.nodes.shape[1] == 3, "Nodes should be 3D coordinates"
+        assert submesh.coord_sys == "cartesian", "Box mesh should be cartesian"
+
+    def test_cylinder_mesh_creation(self):
+        try:
+            from pybamm.meshes.scikit_fem_submeshes_3d import ScikitFemGenerator3D
+        except ImportError:
+            pytest.skip("scikit-fem not available")
+
+        geometry = {
```
Comment (rtimms) [inline tests/unit/test_meshes/test_scikit_fem_submesh_3d.py:52]: would expect these to be vars in cylindrical coords

File: examples/scripts/3d_equation_cylindric.py (hunk 3 above)
Comment (Rishab87) [reply examples/scripts/3d_equation_cylindric.py]: yeah you're right, I'll add support for cylindrical coords

File: examples/scripts/3d_equation_cylindric.py (hunk 4 above)
Comment (Rishab87) [reply examples/scripts/3d_equation_cylindric.py]: we dont need that for 3D. In this 3D implementation, the ScikitFemGenerator3D class uses a quality-based approach. It primarily uses the parameter h (target element edge length) to generate a mesh that meets a certain quality standard. The final number of nodes and elements depends on this, not a direct input.

File: src/pybamm/expression_tree/unary_operators.py (hunk 5 above)
Comment (Rishab87) [reply src/pybamm/expression_tree/unary_operators.py]: left right etc for a box is the face of that geometry only, for cylinder we are using inner radius, outer radius, top and bottom

File: src/pybamm/meshes/scikit_fem_submeshes_3d.py (hunk 6 above)
Comment (Rishab87) [reply src/pybamm/meshes/scikit_fem_submeshes_3d.py:32]: yeah I'll add checks

File: src/pybamm/meshes/scikit_fem_submeshes_3d.py (hunk 7 above)
Comment (Rishab87) [reply src/pybamm/meshes/scikit_fem_submeshes_3d.py]: removing them

File: src/pybamm/expression_tree/unary_operators.py (hunk 5 above)
Comment (Rishab87) [reply src/pybamm/expression_tree/unary_operators.py]: When our mesh generator processes boundary conditions, it dynamically looks for the corresponding boundary facet basis on the mesh object (e.g., getattr(mesh, "inner_radius_basis")).

File: src/pybamm/meshes/scikit_fem_submeshes_3d.py [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+
+
+class ScikitFemGenerator3D(pybamm.MeshGenerator):
+    """
+    A mesh generator that creates 3D tetrahedral meshes using scikit-fem.
+
+    Parameters
+    ----------
+    geom_type : str
+        Type of geometry to generate ('box', 'cylinder', 'spiral')
```
Comment (rtimms) [inline src/pybamm/meshes/scikit_fem_submeshes_3d.py]:
```suggestion
        Type of geometry to generate ('box' or 'cylinder')
```

File: src/pybamm/spatial_methods/scikit_finite_element_3d.py [hunk 11]
```diff
@@ -0,0 +1,624 @@
...
+        pybamm.citations.register("Gustafsson2020")
+
+    def build(self, mesh):
+        super().build(mesh)
+        for dom in mesh.keys():
+            mesh[dom].npts_for_broadcast_to_nodes = mesh[dom].npts
+
+    def spatial_variable(self, symbol):
+        """
+        Creates a discretised spatial variable for x, y, or z.
```
Comment (rtimms) [inline src/pybamm/spatial_methods/scikit_finite_element_3d.py]: or r or theta

File: src/pybamm/spatial_methods/scikit_finite_element_3d.py [hunk 12]
```diff
@@ -0,0 +1,624 @@
...
+            entries = np.sqrt(mesh.nodes[:, 0] ** 2 + mesh.nodes[:, 1] ** 2)[
+                :, np.newaxis
+            ]
+        elif symbol.name == "theta":
+            entries = np.arctan2(mesh.nodes[:, 1], mesh.nodes[:, 0])[
+                :, np.newaxis
+            ]  # pragma: no cover
+        else:
+            raise pybamm.GeometryError(
+                f"Spatial variable must be 'x', 'y' or 'z', not {symbol.name}"
```
Comment (rtimms) [inline src/pybamm/spatial_methods/scikit_finite_element_3d.py]: or r or theta

File: src/pybamm/spatial_methods/scikit_finite_element_3d.py [hunk 13]
```diff
@@ -0,0 +1,624 @@
...
+            )  # pragma: no cover
+        return pybamm.Vector(entries, domains=symbol.domains)
+
+    def gradient(self, symbol, discretised_symbol, boundary_conditions):
+        """
+        Matrix-vector multiplication to implement the 3D gradient operator.
+        Returns a Concatenation of [grad_x, grad_y, grad_z] similar to 2D approach.
```
Comment (rtimms) [inline src/pybamm/spatial_methods/scikit_finite_element_3d.py]: or r, theta. there are a few places in the docstrings that assume cartesian

File: tests/shared.py [hunk 14]
```diff
@@ -116,6 +116,137 @@ def get_mesh_for_testing(
...
+        y = pybamm.SpatialVariable("y", ["current collector"])
+        z = pybamm.SpatialVariable("z", ["current collector"])
+        geometry = {
+            "current collector": {
+                x: {"min": pybamm.Scalar(0), "max": pybamm.Scalar(1)},
+                y: {"min": pybamm.Scalar(0), "max": pybamm.Scalar(1)},
+                z: {"min": pybamm.Scalar(0), "max": pybamm.Scalar(1)},
+            }
+        }
+        var_pts = {x: 5, y: 5, z: 5}
```
Comment (rtimms) [inline tests/shared.py:134]: do we need var_pts here since the mesh generator decides on the number of nodes based on h?

it would be nice if you could grab the `var_pts` after the mesh is created so you know how many nodes you have in x/y/z

File: tests/shared.py [hunk 15]
```diff
@@ -116,6 +116,137 @@ def get_mesh_for_testing(
...
 
 
+def get_unit_3d_mesh_for_testing(geom_type="box", **geom_params):
+    if geom_type == "box":
+        x = pybamm.SpatialVariable("x", ["current collector"])
+        y = pybamm.SpatialVariable("y", ["current collector"])
+        z = pybamm.SpatialVariable("z", ["current collector"])
+        geometry = {
+            "current collector": {
+                x: {"min": pybamm.Scalar(0), "max": pybamm.Scalar(1)},
```
Comment (rtimms) [inline tests/shared.py]: for testing I would make these have different lengths to avoid accidentally covering up errors where an axis gets swapped

File: tests/unit/test_spatial_methods/test_scikit_finite_element_3d.py [hunk 16]
```diff
@@ -0,0 +1,730 @@
...
+
+        model.initial_conditions = {T: pybamm.Scalar(0)}
+        model.variables = {"T": T}
+
+        disc.process_model(model)
+
+        solver = pybamm.AlgebraicSolver()
+        solution = solver.solve(model)
+
+        T_solution = solution.y.flatten()
```
Comment (rtimms) [inline tests/unit/test_spatial_methods/test_scikit_finite_element_3d.py]: can you test against the analytical solution?

File: tests/unit/test_spatial_methods/test_scikit_finite_element_3d.py [hunk 17]
```diff
@@ -0,0 +1,730 @@
...
+        div_grad_u_exact_num = 9 * r * z
+
+        div_grad_u_sym = pybamm.div(pybamm.grad(u))
+        div_grad_u_disc = disc.process_symbol(div_grad_u_sym)
+        div_grad_u_eval = div_grad_u_disc.evaluate(None, u_exact_num).flatten()
+
+        l2_error = np.sqrt(np.mean((div_grad_u_eval - div_grad_u_exact_num) ** 2))
+        assert l2_error < 1.5
+
+    def test_cylinder_mesh_properties(self):
```
Comment (rtimms) [inline tests/unit/test_spatial_methods/test_scikit_finite_element_3d.py:696]: can you add some tests for disctretising known expressions (e.g. y = r * theta + z) and equations?

File: src/pybamm/meshes/scikit_fem_submeshes_3d.py [hunk 18]
```diff
@@ -0,0 +1,311 @@
...
+        mesh = skfem.MeshTet.init_tensor(
+            np.linspace(x_min, x_max, nx),
+            np.linspace(y_min, y_max, ny),
+            np.linspace(z_min, z_max, nz),
+        )
+
+        bnd_facets = mesh.boundary_facets()
+        midpoints = mesh.p[:, mesh.facets[:, bnd_facets]].mean(axis=1)
+
+        boundaries = {
```
Comment (rtimms) [inline src/pybamm/meshes/scikit_fem_submeshes_3d.py:69]: we need to really clearly document these text descriptions of boundaries, or maybe we should call them "x_min", "x_max" etc.?

**Review by rtimms — CHANGES_REQUESTED** [review]

File: src/pybamm/spatial_methods/scikit_finite_element_3d.py (hunk 13 above)
Comment (Rishab87) [reply src/pybamm/spatial_methods/scikit_finite_element_3d.py]: yeah I'll correct all of them

File: tests/shared.py (hunk 14 above)
Comment (Rishab87) [reply tests/shared.py:134]: Yeah we can grab that with .npts, we cant remove var_pts here since it is a compulsory argument in pybamm.Mesh

File: src/pybamm/meshes/scikit_fem_submeshes_3d.py (hunk 18 above)
Comment (Rishab87) [reply src/pybamm/meshes/scikit_fem_submeshes_3d.py:69]: yeah x_min, x_max would be more descriptive I'll change the names

File: examples/scripts/3d_equation_cylindric.py [hunk 19]
```diff
@@ -0,0 +1,176 @@
+import matplotlib.pyplot as plt
```
Comment (rtimms) [inline examples/scripts/3d_equation_cylindric.py]: can you add some comments to this to explain what is going on?

File: examples/scripts/3d_equation_cylindric.py [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+        r: {"min": pybamm.Scalar(R_inner), "max": pybamm.Scalar(R_outer)},
+        z: {"min": pybamm.Scalar(0), "max": pybamm.Scalar(H)},
+    }
+}
+
+submesh_types = {
+    "current collector": ScikitFemGenerator3D(geom_type="cylinder", h=0.08)
+}
+
+var_pts = {r: None, theta: None, z: None}
```
Comment (rtimms) [inline examples/scripts/3d_equation_cylindric.py]: especially here, it would be good to explain more about how the number of elements is chosen in 3D meshes. can we visualise the mesh?

File: src/pybamm/meshes/scikit_fem_submeshes_3d.py [hunk 21]
```diff
@@ -0,0 +1,311 @@
...
+    ----------
+    geom_type : str
+        Type of geometry to generate ('box' or 'cylinder')
+    **gen_params : dict
+        Geometry-specific parameters
+    """
+
+    def __init__(self, geom_type, **gen_params):
+        supported_geometries = ["box", "cylinder"]
+        if geom_type not in supported_geometries:
```
Comment (rtimms) [inline src/pybamm/meshes/scikit_fem_submeshes_3d.py:27]: minor, but can we add a test checking this error is raised?

File: src/pybamm/meshes/scikit_fem_submeshes_3d.py [hunk 22]
```diff
@@ -0,0 +1,311 @@
...
+class ScikitFemGenerator3D(pybamm.MeshGenerator):
+    """
+    A mesh generator that creates 3D tetrahedral meshes using scikit-fem.
+
+    Parameters
+    ----------
+    geom_type : str
+        Type of geometry to generate ('box' or 'cylinder')
+    **gen_params : dict
+        Geometry-specific parameters
```
Comment (rtimms) [inline src/pybamm/meshes/scikit_fem_submeshes_3d.py]: can you document what these are?

File: examples/scripts/3d_equation_cylindric.py [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+solver = pybamm.AlgebraicSolver()
+solution = solver.solve(model)
+
+nodes = mesh["current collector"].nodes
+T_solution = solution.y.flatten()
+
+print(f"Number of FEM nodes: {nodes.shape[0]}")
+print(f"Temperature range: [{T_solution.min():.2f}, {T_solution.max():.2f}]°C")
+
+Nr_vis, Ntheta_vis, Nz_vis = 30, 40, 25
```
Comment (rtimms) [inline examples/scripts/3d_equation_cylindric.py]: let's add some plotting functions in a future PR

File: tests/shared.py (hunk 14 above)
Comment (rtimms) [reply tests/shared.py:134]: OK, I see. We may want to pass `var_pts` anyway since other domains may not be in 3D. Can you add a warning if the submesh type is 3D FEM saying that var_pts will be ignored and the mesh will be created to meet the `h` tolerance. And maybe recommend passing `None` for spatial variables on a 3D domain

File: tests/unit/test_meshes/test_scikit_fem_submesh_3d.py [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+
+
+class TestScikitFemGenerator3D:
+    def test_box_mesh_creation(self):
+        try:
+            from pybamm.meshes.scikit_fem_submeshes_3d import ScikitFemGenerator3D
+        except ImportError:
+            pytest.skip("scikit-fem not available")
+
+        geometry = {
```
Comment (rtimms) [inline tests/unit/test_meshes/test_scikit_fem_submesh_3d.py:14]: can you update these test to have different lengths for each side just to catch any future bugs that flip axes?

File: tests/unit/test_meshes/test_scikit_fem_submesh_3d.py [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+            }
+        }
+
+        mesh_gen = ScikitFemGenerator3D("box", h=0.3)
+
+        mesh = pybamm.Mesh(
+            geometry,
+            {"domain": mesh_gen},
+            {
+                pybamm.standard_spatial_vars.x: 5,
```
Comment (rtimms) [inline tests/unit/test_meshes/test_scikit_fem_submesh_3d.py]: can we just pass a dict with all values set to`None` here?

File: tests/unit/test_meshes/test_scikit_fem_submesh_3d.py [hunk 26]
```diff
@@ -0,0 +1,183 @@
...
+    def test_submesh_properties(self):
+        try:
+            from pybamm.meshes.scikit_fem_submeshes_3d import (
+                ScikitFemGenerator3D,
+                ScikitFemSubMesh3D,
+            )
+        except ImportError:
+            pytest.skip("scikit-fem not available")
+
+        geometry = {
```
Comment (rtimms) [inline tests/unit/test_meshes/test_scikit_fem_submesh_3d.py:99]: see comment about using different lengths

**Review by rtimms — CHANGES_REQUESTED** [review]
@Rishab87 thanks this is looking great! just a few minor comments and then we can merge this and pick up some follow on tasks in a separate PR

File: examples/scripts/3d_equation_cylindric.py (hunk 23 above)
Comment (Rishab87) [reply examples/scripts/3d_equation_cylindric.py]: yesss

File: tests/unit/test_meshes/test_scikit_fem_submesh_3d.py (hunk 25 above)
Comment (Rishab87) [reply tests/unit/test_meshes/test_scikit_fem_submesh_3d.py]: yess

File: tests/unit/test_meshes/test_scikit_fem_submesh_3d.py (hunk 26 above)
Comment (Rishab87) [reply tests/unit/test_meshes/test_scikit_fem_submesh_3d.py:99]: yeah
